sparse-traits/examples/polynomials.rs

```diff
@@ -0,0 +1,138 @@
+pub use sparse_traits::*;
+use std::fmt::Debug;
+
+struct PolynomialSpace;
+impl LinearSpace for PolynomialSpace {
+    type F = f64;
+    type E = Polynomial;
+}
+
+#[derive(Debug)]
+struct Polynomial {
+    monomial_coeffs: Vec<f64>,
+}
+
+impl Polynomial {
+    fn from_monomial(monomial_coeffs: &[f64]) -> Self {
+        Polynomial {
+            monomial_coeffs: monomial_coeffs.to_owned(),
+        }
+    }
+    fn eval(&self, x: f64) -> f64 {
+        self.monomial_coeffs.iter().rev().fold(0., |r, c| r * x + c)
+    }
+}
+
+#[derive(Debug)]
+struct PolynomialView<'a> {
+    monomial_coeffs: &'a [f64],
+}
+struct PolynomialViewMut<'a> {
+    monomial_coeffs: &'a mut [f64],
+}
+
+impl Element for Polynomial {
+    type Space = PolynomialSpace;
+    type View<'a> = PolynomialView<'a> where Self: 'a ;
+    type ViewMut<'a> = PolynomialViewMut<'a> where Self: 'a;
+
+    fn view<'a>(&'a self) -> Self::View<'a> {
+        PolynomialView {
+            monomial_coeffs: &self.monomial_coeffs,
+        }
+    }
+    fn view_mut<'a>(&'a mut self) -> PolynomialViewMut<'a> {
+        PolynomialViewMut {
+            monomial_coeffs: &mut self.monomial_coeffs,
+        }
+    }
+}
+
+struct PointwiseEvaluatorSpace;
+impl LinearSpace for PointwiseEvaluatorSpace {
+    type F = f64;
+    type E = PointwiseEvaluate;
+}
+impl DualSpace for PointwiseEvaluatorSpace {
+    type Space = PolynomialSpace;
+
+    fn dual_pairing(&self, x: &Self::E, p: &<Self::Space as LinearSpace>::E) -> Result<Self::F> {
+        Ok(x.scale * p.eval(x.x))
+    }
+}
+
+struct PointwiseEvaluate {
+    x: f64,
+    scale: f64,
+}
+impl PointwiseEvaluate {
+    fn new(x: f64) -> Self {
+        PointwiseEvaluate { x, scale: 1. }
+    }
+}
+
+impl Element for PointwiseEvaluate {
+    type Space = PointwiseEvaluatorSpace;
+    type View<'a> = &'a PointwiseEvaluate where Self: 'a;
+    type ViewMut<'a> = &'a mut PointwiseEvaluate where Self: 'a;
+    fn view<'a>(&'a self) -> Self::View<'a> {
+        &self
+    }
+    fn view_mut<'a>(&'a mut self) -> Self::ViewMut<'a> {
+        self
+    }
+}
+
+#[derive(Debug)]
+struct Derivative;
+impl OperatorBase for Derivative {
+    type Domain = PolynomialSpace;
+    type Range = PolynomialSpace;
+    fn as_apply(&self) -> Option<&dyn AsApply<Domain = Self::Domain, Range = Self::Range>> {
+        Some(self)
+    }
+}
+impl AsApply for Derivative {
+    fn apply(&self, p: PolynomialView, dp: PolynomialViewMut) -> Result<()> {
+        for (i, c) in p.monomial_coeffs[1..].iter().enumerate() {
+            dp.monomial_coeffs[i] = (1. + i as f64) * c;
+        }
+        dp.monomial_coeffs[p.monomial_coeffs.len() - 1] = 0.;
```

Comment:
There's a bad assumption here that the `dp` slice has the same length, and thus zeroing this one entry (the highest coefficient) is allowed and sufficient. Our views would need to be owned (resizable), which gets in the way of working with subvectors. We could assert compatibility so I'm not too stressed about it.